File: problem20.java
class Solution {
    public boolean isValid(String s) {
        if(s.length() == 1){
            return false;
        }
        char[] pre = new char[s.length()];
        int space = 0;
        for(int i = 0 ; i < s.length(); i++){
            if(s.charAt(i) == '{' || s.charAt(i) == '(' || s.charAt(i) == '['){
                pre[space] = s.charAt(i);
                space++;
            }
            if((s.charAt(i) == '}' && pre[0] == '\0') || 
               (s.charAt(i) == ')' && pre[0] == '\0') || 
               (s.charAt(i) == ']' && pre[0] == '\0')){
                return false;
            }else if((s.charAt(i) == '}' && pre[space-1] != '{') || 
               (s.charAt(i) == ')' && pre[space-1] != '(') || 
               (s.charAt(i) == ']' && pre[space-1] != '[')){
                return false;
            }else if(s.charAt(i) == '}' || s.charAt(i) == ')' || s.charAt(i) == ']'){
                space--;
                pre[space] = '\0';
            }
            System.out.println(pre);
        }
        if(pre[0] != '\0'){
            return false;
        }
        return true;
    }
}

File: problem20.cpp
class Solution {
public:
    bool isValid(string s) {
        if(s.length() == 1){
            return false;
        }
        stack<char> booma;
        map<char, char> pairs;
        pairs.insert(pair<char, char>('}', '{'));
        pairs.insert(pair<char, char>(')', '('));
        pairs.insert(pair<char, char>(']', '['));
        for(int i = 0; i < s.length(); i++){
            if(s[i] == '(' || s[i] == '{' || s[i] == '['){
                booma.push(s[i]);
            }else if(booma.empty() || pairs[s[i]] != booma.top()){
                return false;
            }else {
                booma.pop();
            }
        }
        if(booma.empty()){
            return true;
        }
        return false;
    }
};